### ISSUE TITLE
lws server keep triggering LWS_CALLBACK_ESTABLISHED and cannot send or recieve msgs!!

### ISSUE THREAD

**Issue by Momilijaz96** [warmcat/libwebsockets#1708]
I am writing a simple web socket server using libwebsocket that can exchange data with a client. but my server can send or receive messages and so does my client...please help out what am I doing wrong???? 

**_Websocket_Server.c_**
`
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <libwebsockets.h>

//Called by websocket server every time a event occurs
static int callback_http(
    struct lws_context *context,
    struct lws *wsi,
    enum lws_callback_reasons reason,
    void *user,
    void *in, size_t len
) {
  return 0;
}

static int callback_dumb_increment(struct lws_context * this_context,
                                   struct lws *wsi,
                                   enum lws_callback_reasons reason,
                                   void *user, void *in, size_t len)
{ //reason=LWS_CALLBACK_SERVER_WRITEABLE;
  
  switch (reason)
    {
    
    case LWS_CALLBACK_RECEIVE:   // server recieved data from client
    {
        // create a buffer to hold our response
        // it has to have some pre and post padding. You don't need to care
        // what comes there, libwebsockets will do everything for you. For more info see
         unsigned char *buf = (unsigned char*) malloc((sizeof(char))*(LWS_SEND_BUFFER_PRE_PADDING + 10 +
                                                         LWS_SEND_BUFFER_POST_PADDING));

        int i;

        // pointer to `void *in` holds the incomming request
        // we're just going to put it in reverse order and put it in `buf` with
        // correct offset. `len` holds length of the request.
        printf("Recieved from client:  %s\n",(char *)in);
        if (!in){
          for (i=0; i < len; i++)
          {
              buf[LWS_PRE + (10 - 1) - i ] = ((char *) in)[i];
          } 
        }
        // log what we recieved and what we're going to send as a response.
        // that disco syntax `%.*s` is used to print just a part of our buffer
        printf("received data: %s, replying: %.*s\n", (char *) in, 10,
               buf + LWS_SEND_BUFFER_PRE_PADDING);
      
        free(buf);
        
        //generating WRITABLE callback for sending data
        break;
    }
    case LWS_CALLBACK_SERVER_WRITEABLE:
    {
      /*send data segment from server */
       unsigned char *buf = (unsigned char*) malloc((sizeof(char))*(LWS_PRE + 100 +
                                                         LWS_SEND_BUFFER_POST_PADDING));

        memset(&buf[LWS_PRE], 0, 100);
        printf("Sending data to client: %s\n",buf);
        lws_write(wsi, &buf[LWS_SEND_BUFFER_PRE_PADDING], 100, LWS_WRITE_TEXT);
        free(buf);
        break;
    }
    case LWS_CALLBACK_ESTABLISHED: // just log message that someone is connecting
       {  printf("connection established \n");
        //Request a callback when this socket becomes able to be written to without blocking
          //lws_callback_on_writable(wsi);
        break;
       }

    case LWS_CALLBACK_CLOSED:
        {printf("[Main Service] Client close.\n");
        printf("Websocket inst: %d\n",wsi);
        break;}
    default:
      printf("Callback recieved %d --",reason);
      printf("Unhandled callback \n");
  
       
        break;
    }
    return 0;
}

static struct lws_protocols protocols[] = {
  /* first protocol must always be HTTP handler */
  {
    "http-only",   // name
    callback_http, // callback
    0              // per_session_data_size
  }, 
  {
    "dumb-increment-protocol", // protocol name - very important!
    callback_dumb_increment,   // callback
    0,                          // per session data  
  },
  {
    NULL, NULL, 0   /* End of list */
  }
};

int main(void) {
  // server url will be http://localhost:9000
  int port = 9000;
  const char *interface = NULL;
  struct lws_context_creation_info info;

  struct lws_context *context;
  // we're not using ssl
  const char *cert_path = NULL;
  const char *key_path = NULL;
  // no special options
  int opts = 0;
   memset(&info, 0, sizeof info); /* otherwise uninitialized garbage */
        info.port = port;
        info.iface=NULL;
        info.extensions=NULL;
        info.mounts = NULL;
        info.ssl_cert_filepath=NULL;
        info.ssl_private_key_filepath=NULL;
        info.ssl_ca_filepath=NULL;
        info.gid=-1;
        info.uid=-1;
        info.ka_time=0;
        info.ka_probes=0;
        info.ka_interval=0;
        info.protocols = protocols;
        info.options=0;
  // create libwebsocket context representing this server
  context = lws_create_context(&info);
   
  if (context == NULL) {
    fprintf(stderr, "libwebsocket init failed\n");
    return -1;
  }
   
  printf("starting server...\n");
   
  // infinite loop, to end this server send SIGTERM. (CTRL+C)
  int n=1;
  while (n>=0) {
    n=lws_service(context, 50);
    //printf("Next session : %d\n",n);
    
  }
  free(&info);
  lws_context_destroy(context);
   
  return 0;
}
`

**_Client.html_**
`<!DOCTYPE HTML>

<html>
	<head>
		<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js"></script>
		<script type = "text/javascript">
			
			var socket = new WebSocket('ws://localhost:9000','dumb-increment-protocol');
			//var socket = new WebSocket('ws://echo.websocket.org');	
			// Open the socket
			socket.onopen = function(event) {
				// Send an initial message
				
				setInterval(() => {
				if (socket.bufferedAmount == 0) {
				socket.send("New message");
				console.log("Message sent!!")
				}
				}, 2);
			
				//socket.send('Hello there!');
				console.log("Connection established");
			};

			// Listen for messages
			socket.onmessage = function(event) {
				alert(`[message] Data received from server: ${event.data}`);
				console.log('message recieved from server')
				document.getElementById("name").innerHTML = event.data;
			};

			socket.onerror = function(event){
				alert(`[error] ${error.message}`);
				console.log(error)
			}
	
			// Listen for socket closes
			socket.onclose = function(event) {
				console.log('socket closed', event);
				// To close the socket....
				socket.close()
			};

		</script>
	</head>
	
	<body>
		<p id="name">Please check the console log of your browser</p>
	</body>
</html>
`

**Comment by lws-team** [comment]
Who knows "what you're doing wrong".

If you think it's something to do with lws, show me how to reproduce it using one of the provided minimal examples and I'll take a look at it.

**Comment by Momilijaz96** [comment]
it is a simple websocket server, I am a complete beginner and cannot understand reference from any minimal examples...i just want the server to send messages but it cant!
Kindly point me in the right direction or the minimal example i should look at..Thank u

**Comment by lws-team** [comment]
https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/ws-server/minimal-ws-server

**Comment by Momilijaz96** [comment]
that is an http server .... i have to implement a WebSocket server. What changes to be made in this one? or there's another one?

**Comment by lws-team** [comment]
It's a ws server.

**Comment by Momilijaz96** [comment]
minimal ws server throws this error in console of browser: Refused to load the font '<URL>' because it violates the following Content Security Policy directive: "font-src 'self'".

**Comment by Momilijaz96** [comment]
Hey Andy! I used minimal web server and things are getting much clearer now.. I want my server to send a message but every time the callback switch when comes to LWS_CALLBACK_SERVER_WRITEABLE , it breaks at this condition:
`
if (pss->last == vhd->current){
                 
                        break;}
`
what is this condition actually checking?

**Comment by Momilijaz96** [comment]
I customized it......Thank you for pointing me in right direction.
